pkg/xds/server/callbacks/dataplane_callbacks.go

```diff
@@ -0,0 +1,146 @@
+package callbacks
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	core_model "github.com/kumahq/kuma/pkg/core/resources/model"
+	core_xds "github.com/kumahq/kuma/pkg/core/xds"
+	util_xds "github.com/kumahq/kuma/pkg/util/xds"
+)
+
+// DataplaneCallbacks are XDS callbacks that keep the context of Kuma Dataplane.
+// In the ideal world we could assume that one Dataplane has one xDS stream.
+// Due to race network latencies etc. there might be a situation when one Dataplane has many xDS streams for the short period of time.
+// Those callbacks helps us to deal with such situation.
+//
+// Keep in mind that it does not solve many xDS streams across many instances of the Control Plane.
+// If there are many instances of the Control Plane and Dataplane reconnects, there might be an old stream
+// in one instance of CP and a new stream in a new instance of CP.
+//
+// Those callbacks may be also used with SDS. In case of SDS, at this moment Envoy creates many SDS streams to the Control Plane.
+type DataplaneCallbacks interface {
+	// OnProxyConnected is executed when proxy is connected after it was disconnected before.
+	OnProxyConnected(streamID core_xds.StreamID, dpKey core_model.ResourceKey, ctx context.Context, metadata core_xds.DataplaneMetadata) error
+	// OnProxyReconnected is executed when proxy is already connected, but there is another stream.
+	// This can happen when there is a delay with closing the old connection from the proxy to the control plane.
+	OnProxyReconnected(streamID core_xds.StreamID, dpKey core_model.ResourceKey, ctx context.Context, metadata core_xds.DataplaneMetadata) error
+	// OnProxyDisconnected is executed only when the last stream of the proxy disconnects.
+	OnProxyDisconnected(streamID core_xds.StreamID, dpKey core_model.ResourceKey)
```

Comment:
Should we pass the stream context here too?

Reply:
I'm not sure. Original interface for xds callbacks doesn't have a `context` in `OnStreamClosed`:
```go// Callbacks defines Callbacks for xDS streaming requests. The difference over real go-control-plane Callbacks is that it takes an DiscoveryRequest / DiscoveryResponse interface.
// It helps us to implement Callbacks once for many different versions of Envoy API.
type Callbacks interface {
	// OnStreamOpen is called once an xDS stream is opened with a stream ID and the type URL (or "" for ADS).
	// Returning an error will end processing and close the stream. OnStreamClosed will still be called.
	OnStreamOpen(context.Context, int64, string) error
	// OnStreamClosed is called immediately prior to closing an xDS stream with a stream ID.
	OnStreamClosed(int64)
	// OnStreamRequest is called once a request is received on a stream.
	// Returning an error will end processing and close the stream. OnStreamClosed will still be called.
	OnStreamRequest(int64, DiscoveryRequest) error
	// OnStreamResponse is called immediately prior to sending a response on a stream.
	OnStreamResponse(int64, DiscoveryRequest, DiscoveryResponse)
}
```
I know that we store the `context` in the map for each streamID, but do we really need this? Probably it has `Err()` or `Deadline()` that kind of limits the way of using it.

But that's fairly simple to add, I can do it if needed

Reply:
If there's no current need for it, there's no need to add it in this PR :)